src/connection/mc_worker.erl

```diff
@@ -57,14 +58,62 @@ init(Options) ->
               #state{socket = Socket,
                      conn_state = ConnState,
                      net_module = NetModule,
-                     next_req_fun = NextReqFun});
+                     next_req_fun = NextReqFun}),
+          ignore;
         {error, Reason} ->
-          proc_lib:init_ack({error, Reason})
+          proc_lib:init_ack({error, Reason}),
+          ignore
       end;
     Error ->
-      proc_lib:init_ack(Error)
+      proc_lib:init_ack(Error),
+      ignore
   end.
 
+
+install_mc_worker_info_and_connect(Conf) ->
+    try
+        %% We install info (protocol type) about the worker in an ETS table outside
+        %% the process so we can construct the right kind of messages to send to
+        %% the process
+        ProtocolType =
+            case application:get_env(mongodb, use_legacy_protocol, auto) of
+                true -> legacy;
+                false -> op_msg; %% modern protocol based on the op_msg package
+                auto ->
+                    %% Automatically detect which protocol to use. We send a
+                    %% command using the old protocol. If we get back error code
+                    %% 352* (UnsupportedOpQueryCommand), we are in a version that
+                    %% don't support the legacy protocol so we use the op_msg based
+                    %% protocol instead.
+                    %% * https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml
+                    {ok, InitSocket} = mc_worker_logic:connect_to_database(Conf),
+                    NetModule = get_set_opts_module(Conf),
+                    Database = mc_utils:get_value(database, Conf, <<"admin">>),
+
+                    Command = bson:document([{<<"notExistingCommandXyzErlang">>, <<"void">>}]),
+                    Request = #'query'{
+                                 collection = <<"$cmd">>,
+                                 selector = Command,
+                                 batchsize = -1
+                                },
+                    Response = mc_connection_man:request_raw_no_parse(InitSocket, Database, Request, NetModule),
+                    %% We have to create a new connection because the previous one does not
+                    %% accept new commands after it got a bad command
+                    NetModule:close(InitSocket),
+                    case Response of
+                        [{_, #reply{documents = [#{<<"code">> := 352, <<"ok">> := 0.0}|_]}}|_] ->
+                            op_msg;
+                        _ErrorResponse ->
+                            legacy
+                    end
```

Comment:
move it to another fucntion?

Reply:
Refactored and moved to different module. See commit below.